### DISCUSSION TITLE
Most efficient way to enumerate Ethereum addresses

### DISCUSSION THREAD

**Discussion by sealer3** [erigontech/erigon#6981] [Q&A]
What's the most efficient way to enumerate Ethereum addresses with an Erigon archive node? Ideally I won't have to replay all transactions again. Specifically I am looking for a complete list of contract addresses, but I know I can just check if the address is a contract using `eth_getCode`.

I can't just use the `from` and `to` addresses of transactions because, for instance, a smart contract might have been created and always later called by internal transactions.

I am considering using the RPC call `debug_accountRange` on every block. Is this the right method, correctly returning all accounts, and is there a more efficient way?

**Comment by AskAlexSharov** [comment]
Erigon has no index for filtering contracts. Also has no AccounID to “enumerate”. 

1. @wmitsuda from Otterscan working on something similar
2. You can use TrueBlocks or TheGraph

**Reply by wmitsuda** [reply]
yes, I actually have coded a stage that enumerates contract creation in a temporal order (which is pre-requisite for other stuff/analysis I'm working on), unwind friendly.

That gives you a sorted, paginated view of all contracts created.

I hope to publish a first alpha draft soon.

**Comment by tjayrush** [comment]
Some further context for TrueBlocks:

Introductory: https://trueblocks.io/docs/install/build-unchained-index/ 
Detail: https://trueblocks.io/papers/2023/specification-for-the-unchained-index-v0.51.0-beta.pdf
